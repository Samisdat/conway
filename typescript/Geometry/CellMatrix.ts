import {Position} from '@Conway/Geometry/Position';

interface CellMatrixInterface {
    [index: string]: Position;
}

interface Dimension {
    width: number;
    height: number;
}

export class CellMatrix {

    private matrix: CellMatrixInterface = {};

    public has(position: Position): boolean {

        return (undefined !== this.matrix[position.toString()]);

    }

    public add(position: Position): void {

        if (true === this.has(position)) {
            return;
        }

        this.matrix[position.toString()] = position;

    }

    public remove(position: Position): void {

        if (false === this.has(position)) {
            return;
        }

        delete this.matrix[position.toString()];

    }

    public all(): Position[] {

        const positions: Position[] = [];

        for (const index in this.matrix) {

            positions.push(this.matrix[index]);

        }

        return positions;

    }

    private getDimension(): Dimension {

        const dimension: Dimension = {
            width: 0,
            height: 0
        }

        let left = 0;
        let right = 0;

        let top = 0;
        let bottom = 0;

        for (const position of this.all()) {

            if (position.x < left) {
                left = position.x;
            }

            if (position.x > right) {
                right = position.x;
            }

            if (position.y < bottom) {
                bottom = position.y;
            }

            if (position.y > top) {
                top = position.y;
            }

        }

        if (0 > left && 0 < right) {
            dimension.width = right - left;
        }
        else{
            dimension.width = right - left;
        }

        if (0 > bottom && 0 < top) {
            dimension.height = top - bottom;
        }
        else{
            dimension.height = top - bottom;
        }

        return dimension;

    }

    public width(): number {

        return this.getDimension().width;

    }

    public height(): number {

        return this.getDimension().height;

    }

}
