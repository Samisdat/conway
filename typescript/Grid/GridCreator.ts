import { Position } from '../position';

export class GridCreator {


    public readonly canvasWidth:number;
    public readonly canvasHeigth:number;
    public readonly cellWidthAndHeight:number;
    public readonly pan:Position;

    private rows:number;
    private cols:number;

    constructor(
        canvasWidth:number,
        canvasHeigth:number,
        cellWidthAndHeight:number,
        pan:Position
    ) {
        this.canvasWidth = canvasWidth;
        this.canvasHeigth = canvasHeigth;
        this.cellWidthAndHeight = cellWidthAndHeight;
        this.pan = pan;


        this.rows = this.canvasWidth / cellWidthAndHeight;
        this.cols = this.canvasHeigth / cellWidthAndHeight;

    }

    public getRows():number{
        return this.rows;
    }

    public getCols():number{
        return this.cols;
    }

    public getPan():Position{
        return this.pan;
    }

}