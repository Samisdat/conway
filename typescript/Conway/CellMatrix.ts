import {Cell} from './cell';
import {Position} from './position';

interface CellMatrixInterface {
    [index: string]: Cell;
}

export class CellMatrix {

    private matrix: CellMatrixInterface = {};

    public get(position: Position): Cell | undefined {

        if (undefined !== this.matrix[position.toString()]) {
            return this.matrix[position.toString()];
        }

        return undefined;


    }

    public add(cell: Cell): void {

        if (undefined !== this.matrix[cell.position.toString()]) {
            console.log('argh');
        }

        this.matrix[cell.position.toString()] = cell;

    }

    public remove(cell: Cell): void {

        if (undefined !== this.matrix[cell.position.toString()]) {
            delete this.matrix[cell.position.toString()];
        }

    }

    public all(): Cell[] {

        const cells: Cell[] = [];

        for (const index in this.matrix) {

            cells.push(this.matrix[index]);

        }

        return cells;

    }


}
