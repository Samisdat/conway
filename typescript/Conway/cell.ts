import { Position } from 'Conway/position';

export class Cell {

    public readonly position: Position;

    private nowliving: Boolean = false;

    private thenLiving: Boolean = true;

    constructor(position: Position) {
        this.position = position;
    }

    /**
     * Getter for x allows access like position.x
     */
    get x(): Number {
        return this.position.x;
    }

    get y(): Number {
        return this.position.y;
    }

    isAlive(): Boolean {
        return (true === this.nowliving);
    }

    kill(): void {
        this.thenLiving = false;
    }

    elapse(): void {

        if (true === this.thenLiving) {
            this.nowliving = true;
        }
        else {
            this.nowliving = false;
        }
    }

}