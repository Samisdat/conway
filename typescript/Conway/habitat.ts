import { Position } from './position';
import { Cell } from './cell';
import { LivingCell } from './livingcell';
import { Pattern } from './pattern';
import {Config} from '../Config';

export class Habitat {

    private generationDuration: number;

    private cells: Cell[] = [];

    constructor(generationDuration: number) {

        this.generationDuration = generationDuration;

    }

    public startAging(): void {

        setInterval(
            () => {
                this.elapse();
            },
            this.generationDuration
        );

    }

    private isLiving(position: Position): Boolean {

        let isLiving = false;

        for (let cell of this.cells) {

            if (cell.x === position.x && cell.y === position.y) {

                if (true === cell.isAlive()) {
                    isLiving = true;
                }

                break;
            }
        }

        return isLiving;
    }

    private countLivingNeighbours(position: Position): number {

        let livingNeighbours: number = 0;

        let neighbours: Position[] = position.getNeighbours();

        for (let neighbour of neighbours) {
            let isLiving = this.isLiving(neighbour);

            if (true === isLiving) {
                livingNeighbours += 1;
            }

        }

        return livingNeighbours;

    }

    private applyRules(): void {
        /**
         * there are four rules
         * 1) Any live cell with fewer than two live neighbours dies, as if caused by under-population.
         * 2) Any live cell with two or three live neighbours lives on to the next generation.
         * 3) Any live cell with more than three live neighbours dies, as if by over-population.
         * 4) Any dead cell with exactly three live neighbours becomes a live cell, as if by reproduction.
         *
         *  All rules dealing with living cells can be handled by kill all cells that don't have two or three neighbours
         *  Thats the first loop
         *
         *  To check if any death cells will be birth, we count the living neighbour of all cells neighbours
         *  Thats the second loop
         */

        for (let cell of this.cells) {

            let livingNeighbours = this.countLivingNeighbours(cell.position);

            if (2 === livingNeighbours || 3 === livingNeighbours) {
                continue;
            }

            cell.kill();

        }

        let createCellsAt = {};

        for (let cell of this.cells) {

            let neighbours = cell.position.getNeighbours();

            for (let neighbour of neighbours) {

                let livingNeighbours = this.countLivingNeighbours(neighbour);

                if (3 === livingNeighbours) {
                    createCellsAt[neighbour.x + '-' + neighbour.y] = neighbour;
                }

            }
        }

        for (let pos in createCellsAt) {
            let positionOfNewCell: Position = createCellsAt[pos];
            this.cells.push(new Cell(positionOfNewCell));
        }

    }

    private removeBodies(): void {

        let cleaned: Cell[] = [];

        for (let cell of this.cells) {
            if (true === cell.isAlive()) {
                cleaned.push(cell);
            }
        }

        this.cells = cleaned;

    }

    public elapse(): void {

        this.applyRules();

        for (let cell of this.cells) {
            cell.elapse();
        }

        this.removeBodies();

    }


    public seed(position: Position): void {
        this.cells.push(new LivingCell(position));
    }

    public seedPattern(pattern: Pattern): void {

        const positions = pattern.get();

        for (let position of positions) {
            this.cells.push(new LivingCell(position));
        }

    }

    public getAllCells(): Cell[] {

        return this.cells;

    }

    get(): Position[] {

        let living: Position[] = [];

        for (let cell of this.cells) {
            if (true === cell.isAlive()) {
                living.push(cell.position);
            }
        }

        return living;
    }

}