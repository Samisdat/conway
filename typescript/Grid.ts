import { Position } from './Conway/position';
import { GridCell } from './Grid/GridCell';
import { Habitat } from './Conway/habitat';
import { CellTypeLiving } from 'CellType/CellTypeLiving';
import { CellTypesCenter } from 'CellType/CellTypesCenter';
import { GridDimension } from 'Grid/GridDimension';
import { IntegerPosition } from './Conway/IntegerPosition';
import {CellTypesFactory} from 'CellType/CellTypesFactory';

export class Grid {

    private readonly habitat: Habitat;

    private gridDimension: GridDimension;
    private sourcePosition: IntegerPosition;

    private offset: Position;

    private cells: GridCell[] = [];

    public readonly center: IntegerPosition;

    private absoluteMap: {} = {};

    constructor(
        habitat: Habitat,
        gridDimension: GridDimension,
        sourcePosition: IntegerPosition,
        offset: Position
    ) {

        this.ensureOffsetBetweenMinusOneAndOne(offset);

        this.habitat = habitat;

        this.gridDimension = gridDimension;

        this.sourcePosition = sourcePosition;

        this.offset = offset;

        // create rows

        let relativePointer = new IntegerPosition(0, 0);
        let absolutePointer = this.sourcePosition;

        this.center = relativePointer.clone().move(
            new IntegerPosition(
                Math.floor(this.gridDimension.rows / 2),
                Math.floor(this.gridDimension.cols / 2)
            )
        );

        relativePointer = relativePointer.move(
            new Position(
                -1 * Math.floor(this.gridDimension.rows / 2),
                -1 * Math.floor(this.gridDimension.cols / 2)
            )
        );

        absolutePointer = absolutePointer.move(
            new Position(
                -1 * Math.floor(this.gridDimension.rows / 2),
                -1 * Math.floor(this.gridDimension.cols / 2)
            )
        );

        const moveRight = new Position(
            1,
            0
        );

        // @TODO Use Bounds

        for (let top = 0; top < this.gridDimension.cols; top += 1) {

            for (let right = 0; right < this.gridDimension.rows; right += 1) {

                const gridCell = new GridCell(
                    relativePointer,
                    absolutePointer,
                    this.offset
                );

                const absolutePosition = gridCell.absolutePosition;

                if (undefined === this.absoluteMap[absolutePosition.x]) {
                    this.absoluteMap[absolutePosition.x] = {};
                }

                this.absoluteMap[absolutePosition.x][absolutePosition.y] = this.cells.length;

                this.cells.push(gridCell);

                relativePointer = relativePointer.move(
                    moveRight
                );

                absolutePointer = absolutePointer.move(moveRight);

            }

            relativePointer = relativePointer.move(
                new IntegerPosition(-1 * this.gridDimension.rows, 1),
            );

            absolutePointer = absolutePointer.move(
                new IntegerPosition(-1 * this.gridDimension.rows, 1),
            );

        }

        for (let positionWithLivingCells of habitat.get()) {

            if (undefined === this.absoluteMap[positionWithLivingCells.x]) {
                continue;
            }
            if (undefined === this.absoluteMap[positionWithLivingCells.x][positionWithLivingCells.y]) {
                continue;
            }

            const cellIndex = this.absoluteMap[positionWithLivingCells.x][positionWithLivingCells.y];

            this.cells[cellIndex].setType(CellTypesFactory.get().living());

        }

    }

    private ensureOffsetBetweenMinusOneAndOne(offset: Position) {

        if (1 < Math.abs(offset.x )) {
            throw new Error('offset.x must be between -1 and 1');
        }

        if (1 < Math.abs(offset.y)) {
            throw new Error('offset.y must be between -1 and 1');
        }


    }

    public getSourcePosition(): IntegerPosition {
        return this.sourcePosition;
    }

    public getOffset(): Position {
        return this.offset;
    }

    public getCols(): number {
        return this.gridDimension.cols;
    }

    public getRows(): number {
        return this.gridDimension.rows;
    }

    public getCells() {
        return this.cells;
    }

    public getCell(col: number, row: number): GridCell {

        for (const cell of this.cells) {

            if (cell.relativePosition.x === row && cell.relativePosition.y === col) {
                return cell;
            }
        }

        return undefined;

    }

    public getCellByAbsolutePosition(x: number, y: number): GridCell {

        if (undefined === this.absoluteMap[x]) {
            return undefined;
        }

        if (undefined === this.absoluteMap[x][y]) {
            return undefined;
        }

        const index = this.absoluteMap[x][y];

        return this.cells[index];
    }

}