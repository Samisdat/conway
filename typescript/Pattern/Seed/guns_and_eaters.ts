import {SeedFunction} from '@Conway/Frontend/Web';
import {GridCreator} from '@Conway/Grid/GridCreator';
import {CellMatrix} from '@Conway/Conway/CellMatrix';
import {Position} from '@Conway/Conway/Position';
import {readPatternFromPlainFile} from '@Conway/Pattern/readPatternFromPlainFile';
import {gunsAndEatersPattern} from '@Conway/Pattern/Store/gunsAndEaters.cells';


export const gunsAndEaters: SeedFunction = (gridCreator: GridCreator, matrix: CellMatrix) => {

    const gunsAndEaters = readPatternFromPlainFile(gunsAndEatersPattern);

    const patternGutter = 2;

    let repeat = Math.floor(gridCreator.getRows() / (gunsAndEaters.getWidth() + patternGutter ));

    if (0 === repeat % 2) {
        repeat += 1;
    }

    const patternsPerSide = Math.floor(repeat / 2);

    matrix.seedPattern(
        gunsAndEaters,
        new Position(0, -1)
    );

    let move = new Position(0, -1);

    for (let i = 0; i < patternsPerSide; i += 1) {

        move = move.move(
            new Position(
                -1 * (gunsAndEaters.getWidth() + patternGutter), 0
            )
        );

        matrix.seedPattern(
            gunsAndEaters,
            move
        );

    }

    move = new Position(0, -1);

    for (let i = 0; i < patternsPerSide; i += 1) {

        move = move.move(
            new Position(
                gunsAndEaters.getWidth() + patternGutter, 0
            )
        );

        matrix.seedPattern(
            gunsAndEaters,
            move
        );

    }

};
