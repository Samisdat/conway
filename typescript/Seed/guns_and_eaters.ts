import {GridCreator} from '../Grid/GridCreator';
import {CellMatrix} from '../Conway/CellMatrix';
import {Patterns} from '../Conway/Patterns';
import {Position} from '../Conway/Position';
import {SeedFunction} from '../Frontend/Web';

export const guns_and_eaters: SeedFunction = (gridCreator: GridCreator, matrix: CellMatrix) => {

    const patterns = new Patterns();
    const gunsAndEaters = patterns.get('guns_and_eaters');

    const patternGutter = 2;

    let repeat = Math.floor(gridCreator.getRows() / (gunsAndEaters.getWidth() + patternGutter ));

    if (0 === repeat % 2) {
        repeat += 1;
    }

    let patternsPerSide = Math.floor(repeat / 2);

    matrix.seedPattern(
        patterns.get('guns_and_eaters'),
        new Position(0, -1)
    );


    let move = new Position(0, -1);

    if (0 === repeat % 2) {
        move = move.move(
            new Position(-1 * Math.floor((gunsAndEaters.getWidth() + patternGutter )  / 2 ), 0)
        );
    }

    for (let i = 0; i < patternsPerSide; i += 1) {

        move = move.move(
            new Position(
                -1 * (gunsAndEaters.getWidth() + patternGutter), 0
            )
        );

        matrix.seedPattern(
            patterns.get('guns_and_eaters'),
            move
        );

    }

    move = new Position(0, -1);

    if (0 === repeat % 2) {
        move = move.move(
            new Position(-1 * Math.floor((gunsAndEaters.getWidth() + patternGutter )  / 2 ), 0)
        );
    }

    for (let i = 0; i < patternsPerSide; i += 1) {

        move = move.move(
            new Position(
                gunsAndEaters.getWidth() + patternGutter, 0
            )
        );

        matrix.seedPattern(
            patterns.get('guns_and_eaters'),
            move
        );

    }

};
