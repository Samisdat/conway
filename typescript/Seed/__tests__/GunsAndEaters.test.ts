import {guns_and_eaters} from '../guns_and_eaters';
import {Position} from '../../Conway/Position';
import {CellMatrix} from '../../Conway/CellMatrix';
import {GridCreator} from '../../Grid/GridCreator';

import serializer from '../../../jest-serialize-conway';

describe('guns_and_eaters', () => {

    beforeAll(() => {

        expect.addSnapshotSerializer(serializer);

    });

    test('seed with even grid', () => {

        const gridCreator = new GridCreator(
            200,
            100,
            1,
            new Position(0, 0),
            1
        );

        const matrix = new CellMatrix();

        guns_and_eaters(gridCreator, matrix);

        expect(matrix).toMatchSnapshot();

    });

    test('seed with odd grid', () => {

        const gridCreator = new GridCreator(
            50,
            100,
            1,
            new Position(0, 0),
            1
        );

        const matrix = new CellMatrix();

        guns_and_eaters(gridCreator, matrix);

        expect(matrix).toMatchSnapshot();

    });

});
