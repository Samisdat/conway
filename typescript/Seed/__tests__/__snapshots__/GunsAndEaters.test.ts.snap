// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`guns_and_eaters seed with even grid 1`] = `
OOOOOOOOOOOXOOOOOOOOOOOOOOOOOOOOOOOOOOOOOXXOOOOOOOOOOOOOOOOOXOOOOOOOOOOOOOOOOOOOOOOOOOOOOOXXOOOOOOOOOOOOOOOOOXOOOOOOOOOOOOOOOOOOOOOOOOOOOOOXXOOOOOOOOOOOOOOOOOXOOOOOOOOOOOOOOOOOOOOOOOOOOOOOXXOOOOOOOOOOOOOOOOOXOOOOOOOOOOOOOOOOOOOOOOOOOOOOOXXOOOO
OOOOOOOOOOOXOXOOOOOOOOOOOOOOOOOOOOOOOOOOOXOOOOOOOOOOOOOOOOOOXOXOOOOOOOOOOOOOOOOOOOOOOOOOOOXOOOOOOOOOOOOOOOOOOXOXOOOOOOOOOOOOOOOOOOOOOOOOOOOXOOOOOOOOOOOOOOOOOOXOXOOOOOOOOOOOOOOOOOOOOOOOOOOOXOOOOOOOOOOOOOOOOOOXOXOOOOOOOOOOOOOOOOOOOOOOOOOOOXOOOOO
XXOOOOOOOOOOOOXXOOOOOOXXOOOOOOOOOOOOOOOXOXOOOOOOOXXOOOOOOOOOOOOXXOOOOOOXXOOOOOOOOOOOOOOOXOXOOOOOOOXXOOOOOOOOOOOOXXOOOOOOXXOOOOOOOOOOOOOOOXOXOOOOOOOXXOOOOOOOOOOOOXXOOOOOOXXOOOOOOOOOOOOOOOXOXOOOOOOOXXOOOOOOOOOOOOXXOOOOOOXXOOOOOOOOOOOOOOOXOXOOOOO
XXOOOOOOOOOOOOXXOOOOXOOOXOOOOOOOOOOOOOXOXOOOOOOOOXXOOOOOOOOOOOOXXOOOOXOOOXOOOOOOOOOOOOOXOXOOOOOOOOXXOOOOOOOOOOOOXXOOOOXOOOXOOOOOOOOOOOOOXOXOOOOOOOOXXOOOOOOOOOOOOXXOOOOXOOOXOOOOOOOOOOOOOXOXOOOOOOOOXXOOOOOOOOOOOOXXOOOOXOOOXOOOOOOOOOOOOOXOXOOOOOO
OOOOOOOOOOOOOOXXOOOXOOOOOXOOOOOOOOXXOOOXOOOOOOOOOOOOOOOOOOOOOOOXXOOOXOOOOOXOOOOOOOOXXOOOXOOOOOOOOOOOOOOOOOOOOOOOXXOOOXOOOOOXOOOOOOOOXXOOOXOOOOOOOOOOOOOOOOOOOOOOOXXOOOXOOOOOXOOOOOOOOXXOOOXOOOOOOOOOOOOOOOOOOOOOOOXXOOOXOOOOOXOOOOOOOOXXOOOXOOOOOOO
OOOOOOOOOOOXOXOOOOXXOXOOOXOOOOOOOOXXOOOOOOOOOOOOOOOOOOOOOOOOXOXOOOOXXOXOOOXOOOOOOOOXXOOOOOOOOOOOOOOOOOOOOOOOOXOXOOOOXXOXOOOXOOOOOOOOXXOOOOOOOOOOOOOOOOOOOOOOOOXOXOOOOXXOXOOOXOOOOOOOOXXOOOOOOOOOOOOOOOOOOOOOOOOXOXOOOOXXOXOOOXOOOOOOOOXXOOOOOOOOOOO
OOOOOOOOOOOXOOOOOOOXOOOOOXOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOXOOOOOOOXOOOOOXOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOXOOOOOOOXOOOOOXOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOXOOOOOOOXOOOOOXOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOXOOOOOOOXOOOOOXOOOOOOOOOOOOOOOOOOOOO
OOOOOOOOOOOOOOOOOOOOXOOOXOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOXOOOXOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOXOOOXOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOXOOOXOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOXOOOXOOOOOOOOOOOOOOOOOOOOOO
OOOOOOOOOOOOOOOOOOOOOOXXOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOXXOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOXXOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOXXOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOXXOOOOOOOOOOOOOOOOOOOOOOO
OOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOO
OOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOO
OOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOO
OOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOO
OOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOO
OOOOOOOOOOOOOOOOOOOOOOOXXOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOXXOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOXXOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOXXOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOXXOOOOOOOOOOOOOOOOOOOOOO
OOOOOOOOOOOOOOOOOOOOOOXOOOXOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOXOOOXOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOXOOOXOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOXOOOXOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOXOOOXOOOOOOOOOOOOOOOOOOOO
OOOOOOOOOOOOOOOOOOOOOXOOOOOXOOOOOOOXOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOXOOOOOXOOOOOOOXOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOXOOOOOXOOOOOOOXOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOXOOOOOXOOOOOOOXOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOXOOOOOXOOOOOOOXOOOOOOOOOOO
OOOOOOOOOOOXXOOOOOOOOXOOOXOXXOOOOXOXOOOOOOOOOOOOOOOOOOOOOOOOXXOOOOOOOOXOOOXOXXOOOOXOXOOOOOOOOOOOOOOOOOOOOOOOOXXOOOOOOOOXOOOXOXXOOOOXOXOOOOOOOOOOOOOOOOOOOOOOOOXXOOOOOOOOXOOOXOXXOOOOXOXOOOOOOOOOOOOOOOOOOOOOOOOXXOOOOOOOOXOOOXOXXOOOOXOXOOOOOOOOOOO
OOOOOOOXOOOXXOOOOOOOOXOOOOOXOOOXXOOOOOOOOOOOOOOOOOOOOOOOXOOOXXOOOOOOOOXOOOOOXOOOXXOOOOOOOOOOOOOOOOOOOOOOOXOOOXXOOOOOOOOXOOOOOXOOOXXOOOOOOOOOOOOOOOOOOOOOOOXOOOXXOOOOOOOOXOOOOOXOOOXXOOOOOOOOOOOOOOOOOOOOOOOXOOOXXOOOOOOOOXOOOOOXOOOXXOOOOOOOOOOOOOO
OOOOOOXOXOOOOOOOOOOOOOXOOOXOOOOXXOOOOOOOOOOOOXXOOOOOOOOXOXOOOOOOOOOOOOOXOOOXOOOOXXOOOOOOOOOOOOXXOOOOOOOOXOXOOOOOOOOOOOOOXOOOXOOOOXXOOOOOOOOOOOOXXOOOOOOOOXOXOOOOOOOOOOOOOXOOOXOOOOXXOOOOOOOOOOOOXXOOOOOOOOXOXOOOOOOOOOOOOOXOOOXOOOOXXOOOOOOOOOOOOXX
OOOOOXOXOOOOOOOOOOOOOOOXXOOOOOOXXOOOOOOOOOOOOXXOOOOOOOXOXOOOOOOOOOOOOOOOXXOOOOOOXXOOOOOOOOOOOOXXOOOOOOOXOXOOOOOOOOOOOOOOOXXOOOOOOXXOOOOOOOOOOOOXXOOOOOOOXOXOOOOOOOOOOOOOOOXXOOOOOOXXOOOOOOOOOOOOXXOOOOOOOXOXOOOOOOOOOOOOOOOXXOOOOOOXXOOOOOOOOOOOOXX
OOOOOXOOOOOOOOOOOOOOOOOOOOOOOOOOOXOXOOOOOOOOOOOOOOOOOOXOOOOOOOOOOOOOOOOOOOOOOOOOOOXOXOOOOOOOOOOOOOOOOOOXOOOOOOOOOOOOOOOOOOOOOOOOOOOXOXOOOOOOOOOOOOOOOOOOXOOOOOOOOOOOOOOOOOOOOOOOOOOOXOXOOOOOOOOOOOOOOOOOOXOOOOOOOOOOOOOOOOOOOOOOOOOOOXOXOOOOOOOOOOO
OOOOXXOOOOOOOOOOOOOOOOOOOOOOOOOOOOOXOOOOOOOOOOOOOOOOOXXOOOOOOOOOOOOOOOOOOOOOOOOOOOOOXOOOOOOOOOOOOOOOOOXXOOOOOOOOOOOOOOOOOOOOOOOOOOOOOXOOOOOOOOOOOOOOOOOXXOOOOOOOOOOOOOOOOOOOOOOOOOOOOOXOOOOOOOOOOOOOOOOOXXOOOOOOOOOOOOOOOOOOOOOOOOOOOOOXOOOOOOOOOOO
`;

exports[`guns_and_eaters seed with odd grid 1`] = `
OOOOOOOOOOOXOOOOOOOOOOOOOOOOOOOOOOOOOOOOOXXOOOO
OOOOOOOOOOOXOXOOOOOOOOOOOOOOOOOOOOOOOOOOOXOOOOO
XXOOOOOOOOOOOOXXOOOOOOXXOOOOOOOOOOOOOOOXOXOOOOO
XXOOOOOOOOOOOOXXOOOOXOOOXOOOOOOOOOOOOOXOXOOOOOO
OOOOOOOOOOOOOOXXOOOXOOOOOXOOOOOOOOXXOOOXOOOOOOO
OOOOOOOOOOOXOXOOOOXXOXOOOXOOOOOOOOXXOOOOOOOOOOO
OOOOOOOOOOOXOOOOOOOXOOOOOXOOOOOOOOOOOOOOOOOOOOO
OOOOOOOOOOOOOOOOOOOOXOOOXOOOOOOOOOOOOOOOOOOOOOO
OOOOOOOOOOOOOOOOOOOOOOXXOOOOOOOOOOOOOOOOOOOOOOO
OOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOO
OOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOO
OOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOO
OOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOO
OOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOO
OOOOOOOOOOOOOOOOOOOOOOOXXOOOOOOOOOOOOOOOOOOOOOO
OOOOOOOOOOOOOOOOOOOOOOXOOOXOOOOOOOOOOOOOOOOOOOO
OOOOOOOOOOOOOOOOOOOOOXOOOOOXOOOOOOOXOOOOOOOOOOO
OOOOOOOOOOOXXOOOOOOOOXOOOXOXXOOOOXOXOOOOOOOOOOO
OOOOOOOXOOOXXOOOOOOOOXOOOOOXOOOXXOOOOOOOOOOOOOO
OOOOOOXOXOOOOOOOOOOOOOXOOOXOOOOXXOOOOOOOOOOOOXX
OOOOOXOXOOOOOOOOOOOOOOOXXOOOOOOXXOOOOOOOOOOOOXX
OOOOOXOOOOOOOOOOOOOOOOOOOOOOOOOOOXOXOOOOOOOOOOO
OOOOXXOOOOOOOOOOOOOOOOOOOOOOOOOOOOOXOOOOOOOOOOO
`;
